[DDP Checkpointing] non-reentrant checkpoint tests

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69060

Saved variable hooks checkpointing was added in https://github.com/pytorch/pytorch/pull/69508, this PR adds some tests for DDP.
Specifically, we can support almost all DDP use cases with this new API, such as dynamic module with find_unused_parameters=True. One case remains to be supported, which is static_graph + non-reentrant based checkpointing. The underlying reason this does not work is https://github.com/pytorch/pytorch/issues/58111.

Tests:

- [x] layer checkpointed once with and without reentrant
- [x] layer checkpointed with unused parameters, only for non-reentrant
- [x] Checkpointing the same module twice, only for non-reentrant
- [x] checkpointing twice with static graph, for both reentrant and non-reentrant
- [x] Dynamic module checkpoint, with weight-sharing with non-reentrant
- [ ] DDP static graph checkpoint with non-reentrant

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma committed:
[DDP Checkpointing] Test support for dynamic module checkpointing

With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma commented on torch/utils/checkpoint.py:
```diff
@@ -354,7 +354,11 @@ def inner_unpack(packed):
                     with torch.enable_grad(), torch.cuda.amp.autocast(self.had_autocast_in_fwd):
                         with torch.autograd.graph.saved_tensors_hooks(inner_pack, inner_unpack):
+                            if isinstance(self.function, torch.nn.Module):
+                                self.function._recomputing = True
```
For DDP to work properly with this implementation of checkpointing, it seems that DDP needs to know whether the forward pass is being recomputed or computed for the first time. 

For the current checkpointing support with reentrant autograd, this is a non-issue because some core counting logic in DDP's forward pass is only enabled if `torch.is_grad_enabled()` is True, and one of the fwd passes run by reentrant autograd checkpointing is run under torch.no_grad().

However this implementation does not use torch.no_grad(), hence the checks in DDP's forward pass no longer work. Adding this and checking it wherever `torch.is_grad_enabled()` (as seen in the changes in this PR) makes it work, but not sure if this is a great design.

Any thoughts? @zhaojuanmao @albanD

albanD replied:
This looks a bit hacky to me.
If we thing that users will really need to know this, then we should allow them to give a context manager or hooks that we will call around the re-computation here.

rohan-varma replied:
Sure, I think a context manager supplied by the user could be a good solution.

I do think that we need this as the current checkpointing implementation works for DDP due to running with no_grad, new checkpoint will have issues if there is no way to know about the recomputing.

albanD replied:
Yes, we can add this feature to the existing implementation as well (even though DDP doesn't need it).

rohan-varma replied:
Although, not sure if this is ideal from DDP user perspective. The user would do something like

model = DDP(...)

# forward pass
checkpoint(model, inputs)

and if they needed to pass a context manager into checkpoint(), it would leak some of DDP's internal abstractions. And since checkpoint is called after DDP is applied, there doesn't seem to be a way to have DDP "tell" checkpoint that it needs to be able to distinguish between recomputing or not.

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Checkpointing] Test support for dynamic module checkpointing"


With saved variable hooks based activation checkpointing, DDP can
support activation checkpointing for all use cases, including dynamic modules
with find_unused_parameters=True.

Enhance the tests to test with and without reentrant checkponting. Adds a test
to ensure dynamic module checkpoint works.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

[ghstack-poisoned]

rohan-varma commented:
Removed all changes to checkpoint/DDP APIs, the only issue is with static_graph and non-reentrant checkpointing. @zhaojuanmao

zhaojuanmao reviewed (approved):
awesome tests!  I will address the static graph with new activation checkpointing issue in a separate PR later on. 

I remembered there was an issue when non-static graph works with new activation checkpointing as well, how was that issue resolved?

rohan-varma committed:
Update on "[DDP Checkpointing] non-reentrant checkpoint tests"


Saved variable hooks checkpointing was added in https://github.com/pytorch/pytorch/pull/69508, this PR adds some tests for DDP.
Specifically, we can support almost all DDP use cases with this new API, such as dynamic module with find_unused_parameters=True. One case remains to be supported, which is static_graph + non-reentrant based checkpointing. The underlying reason this does not work is https://github.com/pytorch/pytorch/issues/58111.

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented:
> awesome tests! I will address the static graph with new activation checkpointing issue in a separate PR later on.
> 
> I remembered there was an issue when non-static graph works with new activation checkpointing as well, how was that issue resolved?

@zhaojuanmao I checked this again, there is a difference between:

1) Call checkpoint() within nn.Module that is running inside of DDP
2) Call checkpoint(ddp_model, input) i.e. entire DDP model is checkpointed. 

New autograd APIs work well for (1), not for (2) due to the fact that the DDP forward pass is rerun in (2). The reason we noticed this issue for static_graph is because that is the only test which passes run_checkpoint=True into test_ddp_checkpointing: https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_nccl.py#L2299

rohan-varma committed:
Update on "[DDP Checkpointing] non-reentrant checkpoint tests"


Saved variable hooks checkpointing was added in https://github.com/pytorch/pytorch/pull/69508, this PR adds some tests for DDP.
Specifically, we can support almost all DDP use cases with this new API, such as dynamic module with find_unused_parameters=True. One case remains to be supported, which is static_graph + non-reentrant based checkpointing. The underlying reason this does not work is https://github.com/pytorch/pytorch/issues/58111.

Tests:

- [x] layer checkpointed once with and without reentrant
- [x] layer checkpointed with unused parameters, only for non-reentrant
- [x] Checkpointing the same module twice, only for non-reentrant
- [x] checkpointing twice with static graph, for both reentrant and non-reentrant
- [x] Dynamic module checkpoint, with weight-sharing with non-reentrant
- [ ] DDP static graph checkpoint with non-reentrant

Differential Revision: [D32712126](https://our.internmc.facebook.com/intern/diff/D32712126/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
